docs/window-shims-and-polyfills.md

```diff
@@ -0,0 +1,120 @@
+
+## Location
+
+We have generated a file called `windows-shims-and-polyfiles.ts` in a subfolder of the location where `server.ts is` created.
+
+## How
+
+Because we need to compile your application's code for use in node, we need 3 types of helpers:
+1. TypeScript shims
+2. Node shims
+3. node polyfills
+
+Those are needed even if the code isn't used when your app is rendered during SSR as Universal still needs to read and compile all of the code in your application.
+
+### 1. TypeScript shims
+
+Since many applications will throw type errors while being compiled by Universal, TypeScript shims are needed. If the code inside the app is properly shielded by feature detection and doesn't run during SSR, TS might still complain about a missing type. This is the simplest kind of shim as it only involves the type system. When you look into the provided file, you will see:
+
+```typescript
+declare var global: {
+  /** Reference to the window. */
+  window: Window;
+  /** Reference to the document. */
+  document: Document;
+  // ...
+  /** Reference to SomeLibrary. */
+  SomeLibrary: any;
+  // ...
+};
+```
+
+Lets take `SomeLibrary` as an example. We just assign any to it, so that any reference in the code would not cause an TS compilation error. As we have properly used feature detection, in our application, (as shown below) the code that is actually using the library is _not_ called at all. However, Typescript is still compiling the code, and will error out if it's not available on the global scope.
+
+```typescript
+import { Component, ElementRef, Inject, OnInit, PLATFORM_ID } from '@angular/core';
+
+@Component({
+  selector: 'app-home',
+  templateUrl: './home.component.html',
+  styleUrls: ['./home.component.css'],
+})
+export class HomeComponent implements OnInit {
+  constructor(
+    private elmRef: ElementRef,
+    @Inject(PLATFORM_ID) private platformId: Object
+    ) {}
+
+  ngOnInit(): void {
+    /**
+     * We need the element anyway, and the nativeElement is
+     * only available in the browser, not during SSR!
+     * So we can can use that as feature detection.
+     */
+    const elm = this.elmRef.nativeElement as HTMLElement;
+    if (elm) {
+      /** somehow we made SomeLibrary available in the browser */
+      SomeLibrary.addExitingFunctionality(elm);
+    }
+    /** as an alternative you can check for the existence of the lib */
+    if (SomeLibrary) {
+      SomeLibrary.addExitingFunctionality(elm);
+    }
+    /**
+     * Or use isPlatformBrowser. Don't forget to inject the platformId then.
+     */
+    if (isPlatformBrowser(this.platformId)) {
+      SomeLibrary.addExitingFunctionality(elm);
+    }    
+  }
+}
+```
+
+When you have TypeScript typing error during compilation for SSR, start off with adding the missing type to the global decoration. the type `any` will do, but providing proper typing will help you discover problems sooner.
```

Comment:
When you have a TypeScript error during compilation for SSR, start off with adding the missing type to the global decoration. The `any` type will do, but providing proper typing will help you discover problems sooner.